packages/autocomplete-js/src/__tests__/renderer.test.ts

```diff
@@ -74,17 +89,312 @@ describe('renderer', () => {
+          'div',
+          null,
+          expect.any(Object)
+        );
       },
       renderer: {
         createElement: mockCreateElement,
         Fragment: CustomFragment,
+        render: mockRender,
       },
     });
   });
+
+  test('defaults `render` when not specified in the renderer', () => {
+    const container = document.createElement('div');
+    const panelContainer = document.createElement('div');
+    const CustomFragment = (props: any) => props.children;
+    const mockCreateElement = jest.fn().mockImplementation(preactCreateElement);
+
+    document.body.appendChild(panelContainer);
+
+    autocomplete<{ label: string }>({
+      container,
+      panelContainer,
+      initialState: {
+        isOpen: true,
+      },
+      getSources() {
+        return [
+          {
+            sourceId: 'testSource',
+            getItems() {
+              return [{ label: '1' }];
+            },
+            templates: {
+              item({ item }) {
+                return item.label;
+              },
+            },
+          },
+        ];
+      },
+      render({ createElement, Fragment, render }, root) {
+        expect(render).toBe(preactRender);
+
+        preactRender(createElement(Fragment, null, 'testSource'), root);
+      },
+      renderNoResults({ createElement, Fragment, render }, root) {
+        expect(render).toBe(preactRender);
+
+        preactRender(createElement(Fragment, null, 'testSource'), root);
+      },
+      renderer: {
+        createElement: mockCreateElement,
+        Fragment: CustomFragment,
+      },
+    });
+  });
+
+  test('warns when specifying an incomplete renderer', () => {
+    const container = document.createElement('div');
+    const panelContainer = document.createElement('div');
+    const mockCreateElement = jest.fn().mockImplementation(preactCreateElement);
+    const CustomFragment = (props: any) => props.children;
+    const mockRender = jest.fn().mockImplementation(preactRender);
+
+    document.body.appendChild(panelContainer);
+
+    expect(() => {
+      autocomplete<{ label: string }>({
+        container,
+        panelContainer,
+        initialState: {
+          isOpen: true,
+        },
+        getSources() {
+          return [
+            {
+              sourceId: 'testSource',
+              getItems() {
+                return [{ label: '1' }];
+              },
+              templates: {
+                item({ item }) {
+                  return item.label;
+                },
+              },
+            },
+          ];
+        },
+        renderer: {
+          createElement: mockCreateElement,
+          Fragment: CustomFragment,
+        },
+      });
+    }).toWarnDev(
+      '[Autocomplete] You provided an incomplete `renderer` (missing: `renderer.render`). This can cause rendering issues.' +
+        '\nSee https://www.algolia.com/doc/ui-libraries/autocomplete/api-reference/autocomplete-js/autocomplete/#param-renderer'
+    );
```

Comment:
All users who have already provided `renderer.createElement` and `renderer.Fragment` will get this warning in the next version, although providing it won't change anything because they relied on their own `ReactDOM.render` in `autocomplete.render`.

Thus I think we should also advise them to:

- remove `autocomplete.render` if they were using it just to override `render`
- use the `render` fn coming from `autocomplete.render` to explain that it's where it's used now

Reply:
Yep good call! I can keep this warning for when people provide and incomplete renderer _and_ don't specify the `render` option and have a second one when they do, with suggestions you outlines.

Reply:
Addressed in https://github.com/algolia/autocomplete/pull/920/commits/1d34178c0876b191fd8d3dbb1742dbddd7b9361d.